src/database.rs

```diff
@@ -132,4 +134,99 @@ impl dyn Database {
         let views = self.zalsa().views();
         views.downcaster_for().downcast(self)
     }
+
+    /// Returns information about any Salsa structs.
+    pub fn structs_info(&self) -> Vec<IngredientInfo> {
+        self.zalsa()
+            .ingredients()
+            .filter_map(|ingredient| {
+                let mut size_of_fields = 0;
+                let mut size_of_metadata = 0;
+                let mut instances = 0;
+
+                for slot in ingredient.memory_usage(self)? {
+                    instances += 1;
+                    size_of_fields += slot.size_of_fields;
+                    size_of_metadata += slot.size_of_metadata;
+                }
+
+                Some(IngredientInfo {
+                    count: instances,
+                    size_of_fields,
+                    size_of_metadata,
+                    debug_name: ingredient.debug_name(),
+                })
+            })
+            .collect()
+    }
+
+    /// Returns information about any memoized Salsa queries.
+    ///
+    /// The returned map holds memory usage information for memoized values of a given query, keyed
+    /// by its `(input, output)` type names.
+    pub fn queries_info(&self) -> HashMap<(&'static str, &'static str), IngredientInfo> {
```

Comment:
I don't have a concrete proposal but I was just wondering if we could somehow use `Zalsa::memo_ingredient_indices` or maybe the `jar_map` (by indexing with the erased type id) to get the query ingredient index.